test-framework/core/src/main/java/org/keycloak/testframework/admin/KeycloakAdminClientFactory.java

```diff
@@ -0,0 +1,97 @@
+import org.keycloak.testframework.realm.ManagedRealm;
+import org.keycloak.testframework.realm.ManagedUser;
+
+import java.util.ArrayList;
+import java.util.List;
+
+
+public class KeycloakAdminClientFactory {
+
+    private final List<Keycloak> adminClients = new ArrayList<>();
+
+    private final InstanceContext<KeycloakAdminClientFactory, InjectAdminClientFactory> instanceContext;
+    private final String serverUrl;
+    private final String grantType;
+
+    public KeycloakAdminClientFactory(InstanceContext<KeycloakAdminClientFactory, InjectAdminClientFactory> instanceContext, String serverUrl, String grantType) {
+        this.instanceContext = instanceContext;
+        this.serverUrl = serverUrl;
+        this.grantType = grantType;
+    }
+
+    public Keycloak createMaster() {
+        KeycloakBuilder clientBuilder = KeycloakBuilder.builder()
+                .serverUrl(serverUrl)
+                .grantType(grantType);
+
+        clientBuilder.realm("master").clientId(Config.getAdminClientId()).clientSecret(Config.getAdminClientSecret());
+        Keycloak adminClient = clientBuilder.build();
+        adminClients.add(adminClient);
+        return adminClient;
+    }
+
+    public Keycloak create(String realmRef, String clientId, String user) {
+        ManagedRealm managedRealm = instanceContext.getDependency(ManagedRealm.class, realmRef);
+
+        KeycloakBuilder clientBuilder = createBuilder(managedRealm.getName());
+        RealmResource realmRes = managedRealm.admin();
+
+        setClient(clientBuilder, realmRes, clientId);
+        setUser(clientBuilder, realmRes, user);
+
+        Keycloak adminClient = clientBuilder.build();
+        adminClients.add(adminClient);
+        return adminClient;
+    }
+
+    public Keycloak create(String realmRef, String clientId) {
+        ManagedRealm managedRealm = instanceContext.getDependency(ManagedRealm.class, realmRef);
+
+        KeycloakBuilder clientBuilder = createBuilder(managedRealm.getName());
+        RealmResource realmRes = managedRealm.admin();
+
+        setClient(clientBuilder, realmRes, clientId);
+
+        Keycloak adminClient = clientBuilder.build();
+        adminClients.add(adminClient);
+        return adminClient;
+    }
+
+    private KeycloakBuilder createBuilder(String realm) {
+        return KeycloakBuilder.builder()
+                .serverUrl(serverUrl)
+                .grantType(grantType)
+                .realm(realm);
+    }
+
+    private void setClient(KeycloakBuilder clientBuilder, RealmResource realmRes, String clientId) {
+        ClientRepresentation clientRep = realmRes.clients().findByClientId(clientId).stream()
+                .findFirst().orElseThrow(() -> new TestFrameworkException("Client " + clientId + " not found in managed realm"));
+        clientBuilder.clientId(clientId).clientSecret(clientRep.getSecret());
+    }
+
+    private void setUser(KeycloakBuilder clientBuilder, RealmResource realmRes, String user) {
```

Comment:
See above, this won't work as password is not returned in clear text

Reply:
Does it behave any differently than in the `KeycloakAdminClientSupplier`?
https://github.com/keycloak/keycloak/blob/main/test-framework/core/src/main/java/org/keycloak/testframework/admin/KeycloakAdminClientSupplier.java#L63-L69

Reply:
For injected users, we won't be able to get the password, as it uses the same mechanism